src/ParallelAlgorithms/Amr/Criteria/Random.hpp

```diff
@@ -0,0 +1,82 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <pup.h>
+
+#include "Domain/Amr/Flag.hpp"
+#include "Domain/Structure/ElementId.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace amr::Criteria {
+class Random : public Criterion {
+ public:
+  /// The fraction of the time random refinement does changes the grid
+  struct DoSomethingFraction {
+    using type = double;
+    static constexpr Options::String help = {
+        "The fraction of the time that random refinement will change the "
+        "grid."};
+    static double lower_bound() { return 0.0; }
+    static double upper_bound() { return 1.0; }
+  };
+
+  /// The maximum allowed refinement level
+  struct MaximumRefinementLevel {
+    using type = size_t;
+    static constexpr Options::String help = {
+        "The maximum allowed refinement level."};
+    static size_t upper_bound() { return ElementId<3>::max_refinement_level; }
+  };
+
+  using options = tmpl::list<DoSomethingFraction, MaximumRefinementLevel>;
+
+  static constexpr Options::String help = {"Randomly h-refine the grid"};
+
+  Random() = default;
+
+  Random(const double do_something_fraction,
+         const size_t maximum_refinement_level);
+
+  /// \cond
+  explicit Random(CkMigrateMessage* msg);
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(Random);  // NOLINT
+  /// \endcond
+
+  using compute_tags_for_observation_box = tmpl::list<>;
+
+  using argument_tags = tmpl::list<>;
+
+  template <typename ArrayIndex, typename Metavariables>
+  auto operator()(Parallel::GlobalCache<Metavariables>& /*cache*/,
+                  const ArrayIndex& array_index) const;
+
+  void pup(PUP::er& p) override;
+
+ private:
+  amr::domain::Flag random_flag(size_t current_refinement_level) const;
+
+  double do_something_fraction_{0.0};
+  size_t maximum_refinement_level_{0};
+};
+
+template <typename ArrayIndex, typename Metavariables>
+auto Random::operator()(Parallel::GlobalCache<Metavariables>& /*cache*/,
```

Comment:
do we want criteria to do communication like this? I think there will be some criteria that need to do communication, but that needs 2 steps anyway, so it probably requires something like a `package_data` function first. The associated change is removing the global cache from the call operator(s) (and in the future adding some sending capabalities, but I'd say defer that)